smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/HttpChecksumTraitValidator.java

```diff
@@ -0,0 +1,173 @@
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.neighbor.Walker;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.HttpChecksumProperties;
+import software.amazon.smithy.model.traits.HttpChecksumTrait;
+import software.amazon.smithy.model.traits.HttpHeaderTrait;
+import software.amazon.smithy.model.traits.HttpPrefixHeadersTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * HttpChecksumTraitValidator validates HttpChecksum trait is modeled with at
+ * least one of the request or response property. This also validates the
+ * resolved locations and algorithms list are not empty.
+ */
+@SmithyInternalApi
+public class HttpChecksumTraitValidator extends AbstractValidator {
+
+    private Model model;
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        this.model = model;
+
+        List<ValidationEvent> events = new ArrayList<>();
+        model.shapes(OperationShape.class)
+                .filter(operation -> operation.hasTrait(HttpChecksumTrait.class))
+                .forEach(operation -> events.addAll(validateOperation(operation)));
+
+        return events;
+    }
+
+    private List<ValidationEvent> validateOperation(OperationShape operation) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        HttpChecksumTrait trait = operation.expectTrait(HttpChecksumTrait.class);
+        Optional<HttpChecksumProperties> requestProperty = trait.getRequestProperty();
+        Optional<HttpChecksumProperties> responseProperty = trait.getResponseProperty();
+
+        if (!requestProperty.isPresent() && !responseProperty.isPresent()) {
+            events.add(error(operation, trait,
+                    "The `httpChecksum` trait must have at least one of the `request` or `response` properties set."));
+        }
+
+        requestProperty.ifPresent(property -> {
+            if (property.getLocations().isEmpty()) {
+                events.add(error(operation, trait,
+                        "The `request` property of the `httpChecksum` trait must contain at least one `location` "
+                                + "property entry, found none."));
+            }
+            if (property.getAlgorithms().isEmpty()) {
+                events.add(error(operation, trait,
+                        "The `request` property of the `httpChecksum` trait must contain at least one `algorithms` "
+                                + "property entry, found none."));
+            }
+
+            if (operation.getInput().isPresent()) {
+                Shape shape = model.expectShape(operation.getInput().get());
```

Comment:
The 3 `expectShape` calls that lead to `validatePrefix` calls can have a second parameter of `StructureShape.class` passed so the method returns a `StructureShape`. This allows you to tighten the `validatePrefix` signature from `Shape` to `StructureShape` (see there for more comments.)